Misc refactors

Sija opened the pull request:
SSIA, all of the refactors have been split into separate commits for readability.

Sija committed:
Bump ameba to v0.9

Sija committed:
Move benchmark and usage_example to examples/ dir

Sija committed:
Remove unused Segment(T).type_for method

Sija committed:
Use case where appropriate instead of if/elsif

Sija committed:
Use #<Klass …> instead of Klass(…) in inspect(io)

Sija committed:
Fix bad english

Sija committed:
Rename one-letter variables to sth moar descriptive

Sija committed:
!var.nil? -> !!var

Sija committed:
Simplify RouteSet(T)#add overloads

Sija committed:
Add RoutedResult(T)#[]? delegate

Sija committed:
Use Enumerable#sum instead of manual counting

Sija committed:
Simplify code a bit

Sija committed:
Unescape * in GlobSegment(T) doc comment

Sija committed:
Simplify case branches

Blacksmoke16 commented on src/amber/router/route_set.cr:
```diff
@@ -176,10 +176,11 @@ module Amber::Router
+      case matches.size
+      when 0
         RoutedResult(T).new nil
+      when .> 1
```
Omg, i tried to do something like this before and never thought to add the `.` :neutral_face:

Sija replied:
I reckon we all do have our TIL moments ;)

robacarp commented:
Just as a point of clarification, how much of this is ameba refactor and how much of it isn't?

Sija commented:
@robacarp none of it is.

Sija commented:
@amberframework/contributors @amberframework/core-team Any comments? Is it GTG?

j8r commented on src/amber/router/route_set.cr:
```diff
@@ -69,12 +69,7 @@ module Amber::Router
-    # Add a route to the tree.
-    def add(path, payload : T, constraints : NamedTuple) : Nil
+    def add(path, payload : T, constraints : Hash(Symbol, Regex) | NamedTuple) : Nil
       add_route path, payload, constraints.to_h.transform_keys { |k| k.to_s }
```
I don't think this overload is necessary - can it be removed? The `Symbol`, or whatever type, to `String` conversion can be done upstream.

Sija replied:
Not rly since we need it for `NamedTuple` type, what we might do is to ditch `Hash` from the type restriction but I don't see a problem with keeping it.

j8r replied:
Ok that's fine. In the style side, the line can be reduced by using `transform_keys &.to_s`

Sija committed:
Make RoutedResult Comparable

Sija committed:
Consistenly use #found? within RoutedResult

Sija committed:
Update README

- Moved Travis CI badge into the header
- Added Latest Release badge
- Added “Installation” section
- Fixed usage examples
- Fixed occurrences of outdated path to benchmark.cr

Sija committed:
Use Enumerable#compact_map instead of map+compact

drujensen commented on src/amber/router/routed_result.cr:
```diff
@@ -1,34 +1,32 @@
 
     def found?
-      !@terminal_segment.nil?
+      !!@terminal_segment
```
small bike shedding but consider avoiding `!!`.  I find `.nil?` preferable.

Sija replied:
`!!` is pretty known idiom, personally I find it more readable than `!var.nil?` - double negation split into prefix and postfix making my mind needlessly jump back-and-forth, instead of having it directly upfront.

drujensen replied:
I know.  I see it all the time.  I'm not a fan and consider it a hack from the JS world.  I'm just bike shedding.

robacarp replied:
@drujensen I agree with you.

Sija replied:
Ok, I'll revert it back.

drujensen reviewed (approved):
looks great!

Sija committed:
Give code comments a bit of love

Sija committed:
Make RouteSet#parse_subpaths private

Sija committed:
Reorder method definitions within RouteSet

Sija committed:
Add radix to development_dependencies

Sija committed:
Remove “benchmark” target

Sija committed:
Reorder shard.yml a bit

robacarp commented:
I don't see any problem with this code, but the spirit of this kind of pull request does not appeal to me.

The reason I asked about ameba is, if it's not enforced or recommended by a static analysis tool, it's just a personal preference refactor. If someone else came by with different preferences and refactored all this out, there is no reason to not accept that pull request too. Some call this "bike-shedding", some call it "churn".

Code style is _always_ subjective, there are no rules. Tools like crystal fmt and ameba attempt to provide a standard, but the details are frequently contested.

Sija committed:
Revert "!var.nil? -> !!var"

This reverts commit 60ad3e874b28dad8ce1df359f4e2526bb95f0958.

Sija committed:
Benchmarker refactor

- Remove obsolete getters
- Move amber_router specific tests into the #benchmark_self method
- Rename “router” to “amber_router” for clarity
- Colorize parts of the output

Sija commented:
@robacarp In general I'd agree, yet lots of these were made with readability in mind (what's `s` variable?), some are about pruning obsolete bits of code and docs (confusion factor) or cleaning the API. As I see it, only few are about _code style_ per se, while the rest brings objective improvements to code readability and cohesion.

Sija commented:
ping @amberframework/contributors. If there are no objections I'll go ahead and merge this.

drujensen commented:
@Sija Sorry, I didn't realize this was ready to go.  I have merged.  Thanks for the contrib!
